appcheck/appcheck.go

```diff
@@ -0,0 +1,176 @@
+	jwks, err := keyfunc.Get(JWKSUrl, keyfunc.Options{
+		Ctx:             ctx,
+		RefreshInterval: 6 * time.Hour,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return &Client{
+		projectID: conf.ProjectID,
+		jwks:      jwks,
+	}, nil
+}
+
+// VerifyToken Verifies the given App Check token.
+//
+// VerifyToken considers an App Check token string to be valid if all the following conditions are met:
+//   - The token string is a valid RS256 JWT.
+//   - The JWT contains valid issuer (iss) and audience (aud) claims that match the issuerPrefix
+//     and projectID of the tokenVerifier.
+//   - The JWT contains a valid subject (sub) claim.
+//   - The JWT is not expired, and it has been issued some time in the past.
+//   - The JWT is signed by a Firebase App Check backend server as determined by the keySource.
+//
+// If any of the above conditions are not met, an error is returned. Otherwise a pointer to a
+// decoded App Check token is returned.
+func (c *Client) VerifyToken(token string) (*DecodedAppCheckToken, error) {
+	// References for checks:
+	// https://firebase.googleblog.com/2021/10/protecting-backends-with-app-check.html
+	// https://github.com/firebase/firebase-admin-node/blob/master/src/app-check/token-verifier.ts#L106
+
+	// The standard JWT parser also validates the expiration of the token
+	// so we do not need dedicated code for that.
+	decodedToken, err := jwt.Parse(token, func(t *jwt.Token) (interface{}, error) {
+		if t.Header["alg"] != "RS256" {
+			return nil, ErrIncorrectAlgorithm
+		}
+		if t.Header["typ"] != "JWT" {
+			return nil, ErrTokenType
+		}
+		return c.jwks.Keyfunc(t)
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	claims, ok := decodedToken.Claims.(jwt.MapClaims)
+	if !ok {
+		return nil, ErrTokenClaims
+	}
+
+	rawAud := claims["aud"].([]interface{})
+	aud := []string{}
+	for _, v := range rawAud {
+		aud = append(aud, v.(string))
+	}
+
+	if !contains(aud, "projects/"+c.projectID) {
+		return nil, ErrTokenAudience
+	}
+
+	// We check the prefix to make sure this token was issued
+	// by the Firebase App Check service, but we do not check the
+	// Project Number suffix because the Golang SDK only has project ID.
+	//
+	// This is consistent with the Firebase Admin Node SDK.
+	if !strings.HasPrefix(claims["iss"].(string), appCheckIssuer) {
+		return nil, ErrTokenIssuer
+	}
+
+	if val, ok := claims["sub"].(string); !ok || val == "" {
+		return nil, ErrTokenSubject
+	}
+
+	appCheckToken := DecodedAppCheckToken{
+		Issuer:   claims["iss"].(string),
+		Subject:  claims["sub"].(string),
+		Audience: aud,
+		Expires:  time.Unix(int64(claims["exp"].(float64)), 0),
+		IssuedAt: time.Unix(int64(claims["iat"].(float64)), 0),
+		AppID:    claims["sub"].(string),
+	}
+
+	// Remove all the claims we've already parsed.
+	for _, usedClaim := range []string{"iss", "sub", "aud", "exp", "iat", "sub"} {
+		delete(claims, usedClaim)
+	}
+	appCheckToken.Claims = claims
```

Comment:
Thank you for adding this! 🎉 I forgot to mention this in the feedback. This is great! 🎉 🎉